pandas/io/common.py

```diff
@@ -93,7 +96,9 @@ def _is_url(url):
         return False
 
 
-def _expand_user(filepath_or_buffer):
+def _expand_user(
+    filepath_or_buffer: FilePathOrBuffer[AnyStr]
```

Comment:
What does subscripting with `AnyStr` do here? Add to Union?

Reply:
FilePathOrBuffer is a just a Union. FilePathOrBuffer[AnyStr] with subscription is effectively a TypeVar in IO. so IO[str] can't become IO[bytes].

i think https://mypy.readthedocs.io/en/latest/generics.html#generic-type-aliases explains it. will send a different link if i come across a better explanation.

Reply:
That's pretty cool - thanks for sharing! Does this really change anything though? We already have `IO[AnyStr]` in `FilePathOrBuffer` so this just restates that (?)

Unrelated note - `IO[AnyStr]` might itself be wrong as `AnyStr` is a TypeVar and I think we need to parametrize `IO` with the actual type. I find the Python docs rather confusing on that so maybe we remove `AnyStr` altogether but can level that as a separate exercise (unless it helps simplify annotation here)

Reply:
> Unrelated note - `IO[AnyStr]` might itself be wrong as `AnyStr` is a TypeVar and I think we need to parametrize `IO` with the actual type.

The alias is a Union, the Union has one and only one Generic. so parametrising the alias is parametrising the only Generic in the Union, i.e. IO. 


> We already have `IO[AnyStr]` in `FilePathOrBuffer` so this just restates that (?)

I don't think AnyStr is treated as a TypeVar inside the union when the alias is defined. so that's why it's needed in use


> Does this really change anything though?

yes. mypy will fail without it.

in to_html etc only string buffers are supported. hence `Optional[FilePathOrBuffer[str]]` is used (note the parametrisation of FilePathOrBuffer here)  and the TypeVar then becomes necessary otherwise a bytes buffer could be returned.

`buffer_put_lines(buf: IO[str], lines: List[str]) -> None` only supports string buffers. so mypy will raise if we don't use TypeVars to maintain the FilePathOrBuffer type in and out of the common functions.

Reply:
Right I think the std documentation isn't very clear but just defining `IO` creates a generic accepting a type of `AnyStr` which is str / bytes. This is in contrast to other generics that really accept type `T` (essentially anything). You can see this if you try to inject a non-str or bytes type

```python
from typing import IO

foo: IO[int]
```

yields

```sh
error: Value of type variable "AnyStr" of "IO" cannot be "int"
```

So I think an error to keep re-parametrizing IO with `AnyStr` in the _typing module and here. 

Let's leave to a follow up to clean up though